include/AST/Types/EnumTy.hpp

```diff
@@ -2,60 +2,98 @@
 #define GLU_AST_TYPES_ENUMTY_HPP
 
 #include "Basic/SourceLocation.hpp"
-#include "TypeBase.hpp"
+#include "Types/TypeBase.hpp"
 
+#include <cassert>
 #include <llvm/ADT/APInt.h>
-#include <llvm/ADT/STLExtras.h>
 #include <llvm/ADT/SmallVector.h>
 #include <llvm/ADT/StringRef.h>
+#include <llvm/Support/Allocator.h>
+#include <llvm/Support/TrailingObjects.h>
 #include <string>
+#include <vector>
 
 namespace glu::types {
 
-/// @brief EnumTy is a class that represents enumerations declared in code.
-class EnumTy : public TypeBase {
+struct Case {
+    std::string name;
+    llvm::APInt value;
+
+    Case(std::string const &n, llvm::APInt const &v) : name(n), value(v) { }
+};
+
+class EnumTy final : public TypeBase,
+                     private llvm::TrailingObjects<EnumTy, Case> {
 public:
-    struct Case {
-        std::string name;
-        llvm::APInt value;
-    };
+    friend llvm::TrailingObjects<EnumTy, Case>;
 
 private:
     std::string _name;
-    llvm::SmallVector<Case> _cases;
+    unsigned _numCases;
     SourceLocation _definitionLocation;
 
+    // Method required by llvm::TrailingObjects to determine the number
+    // of trailing objects.
+    size_t
+        numTrailingObjects(typename llvm::TrailingObjects<EnumTy, Case>::OverloadToken<Case>) const
+    {
+        return _numCases;
+    }
+
+    EnumTy(std::string name, unsigned numCases, SourceLocation loc)
+        : TypeBase(TypeKind::EnumTyKind)
+        , _name(std::move(name))
+        , _numCases(numCases)
+        , _definitionLocation(loc)
+    {
+    }
+
 public:
-    /// @brief Constructor for the EnumTy class.
     EnumTy(
-        std::string name, llvm::SmallVector<Case> cases,
-        SourceLocation definitionLocation
+        std::string const &name, llvm::SmallVectorImpl<Case> const &cases,
+        SourceLocation loc
     )
-        : TypeBase(TypeKind::EnumTyKind)
-        , _name(name)
-        , _cases(cases)
-        , _definitionLocation(definitionLocation)
+        : EnumTy(name, cases.size(), loc)
     {
     }
 
-    std::string getName() const { return _name; }
+    /// @brief Creates an EnumTy by allocating a memory block containing the
+    /// object and its trailing objects for the cases.
+    ///
+    /// @param allocator The allocator (for example, llvm::BumpPtrAllocator)
+    /// used for allocation.
+    /// @param name The name of the enumeration.
+    /// @param cases The cases to copy into the trailing storage.
+    /// @param loc The definition location.
+    /// @return A pointer to the new EnumTy.
+    static EnumTy *create(
+        llvm::BumpPtrAllocator &allocator, std::string const &name,
+        llvm::SmallVectorImpl<Case> const &cases, SourceLocation loc
```

Comment:
if you're making a copy anyway, take an llvm::ArrayRef
```suggestion
        llvm::ArrayRef<Case> const &cases, SourceLocation loc
```